### PR TITLE
[ffmpeg] Warn/fail when selecting unsupported features.

### LINKED ISSUE #17136: [ffmpeg] build failure
**Host Environment**
- OS: Windows
- Compiler: MSVC 14.2

**To Reproduce**
`./vcpkg install ffmpeg[tesseract]:x64-windows-static`

**Failure logs** 
[build-x64-windows-static-rel-out.log](https://github.com/microsoft/vcpkg/files/6273901/build-x64-windows-static-rel-out.log)
[config.log](https://github.com/microsoft/vcpkg/files/6273898/config.log)

**Additional context**
ffmpeg is unable to be built with tesseract after latest PR https://github.com/microsoft/vcpkg/pull/15787

### PR DESCRIPTION
**Pull request by Sibras**
Fixes #17136 #17139

### CODE REVIEW

**Review by JackBoosY — APPROVED** [review]

**Review by ytbigvinny27 — APPROVED** [review]

**Comment by JackBoosY** [comment]
Wow... what's wrong here?

**Comment by Sibras** [comment]
> 
> 
> Wow... what's wrong here?

Not sure what happened with x64_windows as that seems to have a whole pile of errors in ports not even touched by this PR. My only conclusion is that it must be something going funky on the CI. Hopefully if it just gets forced to re-run that might fix it.

**Comment by JackBoosY** [comment]
It's okay now.

**Comment by autoantwort** [comment]
Is is not possible to set the supports field of a feature instead?

**Comment by JackBoosY** [comment]
Ping @strega-nil for review and merge this PR.

**Comment by BillyONeal** [comment]
Tagging with requires:discussion to ask whether having features that silently disable themselves when inapplicable is ok

**Comment by BillyONeal** [comment]
> Tagging with requires:discussion to ask whether having features that silently disable themselves when inapplicable is ok

I misread this and have feedback will be posted shortly.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
Improving error reporting for missing features is a good thing, and I do appreciate that part of this PR. Unfortunately I don't think the "make those errors go away if all is selected" part is reasonable. This causes unfortunate side effects like:

`vcpkg install a-port-wanting-fdk-aac:arm-windows`

fails because that's unsupported but

`vcpkg install a-port-wanting-fdk-aac:arm-windows ffmpeg[all]:arm-windows`

succeeds "for some reason".

As a result I think we have to make those always FATAL_ERROR.

There is a "feature ask" for the vcpkg tool itself to be able to mark features with supports expressions and warn users early if they picked something that won't work of course, which would be a good add to make that big "all" feature definition easier.

**Comment by Sibras** [comment]
> `vcpkg install a-port-wanting-fdk-aac:arm-windows ffmpeg[all]:arm-windows`
> 
> succeeds "for some reason".

Ive updated this so that now all-gpl/all-nonfree are checked so that things like fdk will now fatal error in the "all" case. In the "all-nonfree" case it will emit a warning but it removes fdk from the feature list. Im not sure how else to signal to vcpkg that fdk is not in the feature list and that "a-port-wanting-fdk-aac" should fail.

> There is a "feature ask" for the vcpkg tool itself to be able to mark features with supports expressions and warn users early if they picked something that won't work of course, which would be a good add to make that big "all" feature definition easier.

This would definitely be the better solution. And I see this PR being a stop gap until something like that is implemented.

In the meantime I can revert to having the feature checks back in the CONTROL file under "all" etc. and then have the standalone error checking in the portfile. I was just trying to avoid duplication and the potential for the 2 files to get out of sync. Let me know which is preferred.

**Comment by JackBoosY** [comment]
Ping @BillyONeal for review again.

**Comment by JackBoosY** [comment]
Can you please resolve the file conflicts?

Thanks.

File: versions/f-/ffmpeg.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "0f4a72e4c5d1db1487fe8d33fd8aaece95cba5a3",
```
Comment (JackBoosY) [inline versions/f-/ffmpeg.json:4]:
```suggestion
      "git-tree": "3d6c5743b5652afeeb19816aaf2028a5ab3e465b",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
Waiting for https://github.com/microsoft/vcpkg/pull/18116 merge.

**Review by JackBoosY — APPROVED** [review]

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
There are a few things that need to be done before this is mergeable:

- First, we should switch to using `VCPKG_TARGET_IS_UWP` or `VCPKG_TARGET_IS_WINDOWS`, rather than checking `VCPKG_CMAKE_SYSTEM_NAME`
- Second, please use `vcpkg_fail_port_install` for at least the simpler checks
- Third, remove the `list(REMOVE_ITEM`s

**Comment by JackBoosY** [comment]
@strega-nil @strega-nil-ms can you please review again?

Thanks.

**Review by JackBoosY — APPROVED** [review]

**Comment by ras0219-msft** [comment]
This seems like a good improvement to me, thanks for the PR!

**Pull request merged**